### PR TITLE
Automatically create metaschemas when accessing JSONSchema.metaschema

### LINKED ISSUE #68: "$schema" access should load a Metaschema like "$ref" access loads a JSONSchema
json-schema-org/JSON-Schema-Test-Suite#646 tests that unrecognized optional vocabularies are allowed, which requires treating a schema encountered through `"$schema"` as a meta-schema whether it was explicitly created that way or not.

In [§9.3.1 Detecting a Meta-Schema](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-detecting-a-meta-schema), the spec makes it clear that use of a URI in `"$schema"` requires treating the referenced schema as a meta-schema, even if it was otherwise considered a normal schema:

> Implementations MUST recognize a schema as a meta-schema if it is being examined because it was identified as such by another schema's "$schema" keyword. This means that a single schema document might sometimes be considered a regular schema, and other times be considered a meta-schema.
> 
> In the case of examining a schema which is its own meta-schema, when an implementation begins processing it as a regular schema, it is processed under those rules. However, when loaded a second time as a result of checking its own "$schema" value, it is treated as a meta-schema. So the same document is processed both ways in the course of one session.
> 
> Implementations MAY allow a schema to be explicitly passed as a meta-schema, for implementation-specific purposes, such as pre-loading a commonly used meta-schema and checking its vocabulary support requirements up front. Meta-schema authors MUST NOT expect such features to be interoperable across implementations.

The current `jschon.jsonschema.JSONSchema.metaschema` property code assumes the MAY scenario in the 3rd paragraph by asserting that it is already loaded as a meta-schema.  Instead, the catalog needs to offer a parallel to `get_schema` for metaschemas, which the `JSONSchema.metaschema` property would use.

While the metaschema for a `JSONSchema` instance can be set explicitly as well as detected via `"$schema"`, explicit setting of a meta-schema URI is also sufficient to require interpreting it as a meta-scheme whether it was initially loaded that way or not.

### PR DESCRIPTION
**Pull request by handrews**
Fixes #68 (2nd try, now with much more testing!) , although I'm submitting it as a draft to indicate that I'm not particularly confident that this is the ideal way to solve it. But it's a starting point, at least.

This change allows the tests added in PR json-schema-org/JSON-Schema-Test-Suite#646 (updated here in PR #67) to pass, in accordance with §9.3.1 of the spec, "Detecting a Meta-Schema".

* `Catalog.get_metaschema()` is added, and is analogous in behavior to `Catalog.get_schema()`, checking the `__meta__` cache directly and then relying on `Catalog.create_metaschema()` to avoid needlessly creating both `JSONSchema` and `Metaschema` instances.
* It is called from the now-cached `JSONSchema.metaschema` property, to keep the parallel with references which are only resolved when used.
* `Catalog.create_metaschema()` now returns the created metaschema to avoid having to immediately look it up again, as does `Catalog.create_vocabulary()`
* The `core_vocabulary` parameters have become `default_core_vocabulary` parameters to allow not knowing the core in advance
*The `Metaschema` constructor now looks in `"$vocabulary"` for a vocabulary URI matching `r'^https://json-schema\.org/draft/[^/]*/core$` and if it finds a unique match, uses that instead of the default
* Lack of a recognizable core vocabulary still results in a JSONSchemaError exception if no default is provided
* I wasn't entirely sure where to put the `Metaschema` test cases, so the `Catalog`-related ones went into that test file and the constructor ones went into a new `test_metaschema.py`

Note that the fixture is a separate commit because it is shared with the next PR I'm about to post.

### CODE REVIEW

**Comment by handrews** [comment]
Having spent a great deal more time in all the relevant areas of the code, I haven't come up with any better approach, so I've moved this out of draft PR status.

**Comment by marksparkza** [comment]
Nice work, @handrews! Do you mind if I push to this branch?

**Comment by handrews** [comment]
@marksparkza Thanks!  I think I opened this PR with the "allow maintainer to make changes"  - thanks for checking but feel free to add commits.  What workflow do you have in mind?  Are you looking to discuss those additional commits?  (Just trying to understand the difference between what you want to do here and merging and then adding things separately).

**Comment by handrews** [comment]
Or did you mean create this branch in your repo rather than going straight to main?  I'm also totally fine with that.

**Comment by marksparkza** [comment]
Automated meta-schema creation is a big win, and I think it's worth rounding off this PR by removing boilerplate `create_metaschema()` calls in catalog setup scripts, and allowing `Metaschema` instances to be created implicitly. I realised while working through your changes that this could be done with just a couple of tweaks, so I've added two commits locally: one to update the test suite (identical to the one on the testsuite-main branch, which I'll drop), and one to remove boilerplate.

There are one or two changes that I'd still like to take a closer look at, so I may add further comments following the additional commits. Likewise, please feel free to comment on or query any changes that I make to your code.

Once this PR is merged we're in a good position to publish a very long awaited v0.10, so I'm going to go ahead with that and work through the remaining PRs for the next version increment.

File: CHANGELOG.rst
Comment (marksparkza) [inline CHANGELOG.rst:33]: This can also go under miscellaneous. The params are declared as positional args, so any existing usage of these methods that follow the API and the examples won't need changing.

File: jschon/jsonschema.py
Comment (marksparkza) [inline jschon/jsonschema.py:224]: This line is no longer needed.

File: jschon/jsonschema.py
Comment (marksparkza) [inline jschon/jsonschema.py:221]: This makes sense. I think it implies that the `metaschema_uri` getter should also be cached.

**Review by marksparkza — CHANGES_REQUESTED** [review]

**Comment by handrews** [comment]
@marksparkza Sounds great!  I will try to finish the PR for the `list` output format by Monday as I'd like to get that in if possible, but that's the last thing I really wanted to contribute to 0.10.

I have some other things in the works for draft-next support, but that can wait (and TBH I'm about to not have as much time avaialble so they'd probably take a while before I got around to them).

File: jschon/jsonschema.py
Comment (handrews) [reply jschon/jsonschema.py:221]: It turns out that `metaschema_uri` and `uri` need to be set by `"$schema"` and `"$id"` and that's why they're not cached.  Ideally, they could only be set once and cached after that.  It's still safe to cached `metaschema` because if it has not been set then it will raise an exception, and it really shouldn't ever be changed.

**Comment by handrews** [comment]
> Once this PR is merged we're in a good position to publish a very long awaited v0.10, so I'm going to go ahead with that and work through the remaining PRs for the next version increment.

@marksparkza I think I slightly misunderstood you although it doesn't really change anything.  Just to be sure I understand:  You plan to merge this PR once it's finished and immediately release 0.10.0, correct?

If so, I would recommend also merging the _very simple_  PRs #72 (putting the not-found schema file name in the `CatalogError` message so it's not so cryptic) and #73 (a few super-simple tests for the `Catalog`).  Both were originally part of this PR that I split out to keep this one more focused.  Neither should take long to review.

I'm perfectly happy with the rest (including the `list` output format that I haven't posted yet) going into 0.11.0 (not that I'd try to veto your package's release schedule anyway, even over the above two PR recommendations!).  Would you consider those remaining PRs enough to release 0.11.0 whenever you get through them, or would there be more that you'd want to have in that release?

**Comment by marksparkza** [comment]
> plan to merge this PR once it's finished and immediately release 0.10.0, correct?

Yes, that's the plan. I'd prefer to merge this PR as is, and subsequently look at 72 and 73 for a 0.10.1 patch. I do appreciate that they are quite straightforward, but they're also relatively insignificant and shouldn't hold up or force a rebase of 71. Having said that, we're looking at a timescale of days rather than months for 0.10.1.

> when you push the 0.10.0 release, I'll rebase the other PRs

Thanks, much appreciated. Please just do 72 and 73, and the rest once 0.10.1 is out. I'll be happy to include the `list` PR in 0.10.1 if you're ready with it. Otherwise, nothing else goes into 0.10.1.

>  Would you consider those remaining PRs enough to release 0.11.0

Yes, absolutely!

File: jschon/jsonschema.py
Comment (marksparkza) [reply jschon/jsonschema.py:221]: Right, I see that `@cached_property` doesn't work with a setter. At the time that I modified the `parentschema` property to be cached (as well as `path` and `value` on `JSON`), I didn't do the same for `metaschema` precisely because `metaschema_uri` is updatable. Previously I've vaguely considered trying to resolve this by integrating core keywords (or at least, the 'bootstrapping' ones like `$schema` and `$id`) more tightly with `JSONSchema`, and making `JSONSchema` immutable. But that would really be overly restrictive, and in fact it was my own extension [jschon-translation](https://github.com/marksparkza/jschon-translation) that roundly dismissed the idea of making `JSONSchema` immutable. In any case, in normal usage `uri` and `metaschema_uri` are and should only be set once, and it makes sense to cache `metaschema` because it's accessed repeatedly.

**Pull request merged**
